benchmark/runner/mini_runners.cpp

```diff
@@ -0,0 +1,1161 @@
+    brain::ExecutionOperatingUnitFeatureVector pipe1_vec;
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::SEQ_SCAN, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::SORT_BUILD, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::SORT_ITERATE, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    units.RecordOperatingUnit(execution::pipeline_id_t(0), std::move(pipe0_vec));
+    units.RecordOperatingUnit(execution::pipeline_id_t(1), std::move(pipe1_vec));
+
+    std::stringstream cols;
+    for (auto i = 1; i <= num_col; i++) {
+      cols << "col" << i;
+      if (i != num_col) {
+        cols << ", ";
+      }
+    }
+
+    std::stringstream tbl_name;
+    tbl_name << "SELECT " << (cols.str()) << " FROM INTEGERCol15Row" << row << "Car" << car << " ORDER BY "
+             << (cols.str());
+    BenchmarkSqlStatement(qid, tbl_name.str(), &units, std::make_unique<optimizer::TrivialCostModel>(), true);
+    metrics_manager_->Aggregate();
+    metrics_manager_->UnregisterThread();
+  }
+
+  state.SetItemsProcessed(row);
+}
+
+/**
+ * Arg: <0, 1, 2>
+ * 0 - Number of columns to select
+ * 1 - Number of rows
+ * 2 - Cardinality
+ */
+BENCHMARK_REGISTER_F(MiniRunners, SortRunners)->Unit(benchmark::kMillisecond)->Iterations(1)->Apply(GenScanArguments);
+
+// NOLINTNEXTLINE
+BENCHMARK_DEFINE_F(MiniRunners, HashJoinRunners)(benchmark::State &state) {
+  auto num_col = state.range(0);
+  auto row = state.range(1);
+  auto car = state.range(2);
+
+  // NOLINTNEXTLINE
+  for (auto _ : state) {
+    metrics_manager_->RegisterThread();
+
+    execution::query_id_t qid;
+    switch (num_col) {
+      case 1:
+        qid = HJ_COL_1_QID;
+        break;
+      case 4:
+        qid = HJ_COL_4_QID;
+        break;
+      case 8:
+        qid = HJ_COL_8_QID;
+        break;
+      case 15:
+        qid = HJ_COL_15_QID;
+        break;
+      default:
+        UNREACHABLE("Unexpected number of columns for HJ");
+        break;
+    }
+
+    brain::PipelineOperatingUnits units;
+    brain::ExecutionOperatingUnitFeatureVector pipe0_vec;
+    brain::ExecutionOperatingUnitFeatureVector pipe1_vec;
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::SEQ_SCAN, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::HASHJOIN_BUILD, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::SEQ_SCAN, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::HASHJOIN_PROBE, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
```

Comment:
Unlike other operating units in the execution engine, the `car` here should be the number of rows matched by the hash table look up from the probe table. You can refer [here](https://github.com/malin1993ml/terrier/blob/modeling-metrics/script/model/execution_engine_runner/generate_hashjoin.py#L185) on how I get this number. Note that in that script I hacked the hash join key comparison function so that the hash join only uses the value of the last column. You might need to change all the column `Dist` types to `Rotate` in the `TableGenerator`, and make every column to have the same value (maybe by fixing the random seed)?

You may also want to limit the total cardinality here by controlling the input argument for this runner, like what I did [here](https://github.com/malin1993ml/terrier/blob/modeling-metrics/script/model/execution_engine_runner/generate_hashjoin.py#L189). Otherwise the matched tuple can explode. Similar to the `NLJoin`. Also be careful there because every pair of tuples are going to be "matched" in a nested loop join.

We can meet to discuss this if it's unclear.

Reply:
Changed this such that `col15` is used multiple times.